4.0 multi-db concurrency issue at startup

Seems to be a race condition at startup

possibly use CHM here: https://github.com/neo4j-contrib/neo4j-apoc-procedures/blob/4.0/src/main/java/apoc/RegisterComponentFactory.java#L41

```
2020-01-22 12:04:45.208+0000 WARN [o.n.k.d.Database] [db2] Exception occurred while starting the database. Trying to stop already started components. Component 'apoc.ApocExtensionFactory$ApocLifecycle@3b02b9fb' failed to initialize. Please see the attached cause exception "null".
org.neo4j.kernel.lifecycle.LifecycleException: Component 'apoc.ApocExtensionFactory$ApocLifecycle@3b02b9fb' failed to initialize. Please see the attached cause exception "null".
        at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.init(LifeSupport.java:426)
        at org.neo4j.kernel.lifecycle.LifeSupport.init(LifeSupport.java:66)
        at org.neo4j.kernel.extension.AbstractExtensions.init(AbstractExtensions.java:76)
        at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.init(LifeSupport.java:405)
        at org.neo4j.kernel.lifecycle.LifeSupport.init(LifeSupport.java:66)
        at org.neo4j.kernel.database.Database.initializeExtensions(Database.java:499)
        at org.neo4j.kernel.database.Database.start(Database.java:347)
        at org.neo4j.dbms.database.AbstractDatabaseManager.startDatabase(AbstractDatabaseManager.java:187)
        at com.neo4j.dbms.database.MultiDatabaseManager.forSingleDatabase(MultiDatabaseManager.java:112)
        at com.neo4j.dbms.database.MultiDatabaseManager.startDatabase(MultiDatabaseManager.java:98)
        at com.neo4j.dbms.DbmsReconciler.start(DbmsReconciler.java:537)
        at com.neo4j.dbms.Transitions$TransitionFunction.lambda$prepare$0(Transitions.java:219)
        at com.neo4j.dbms.DbmsReconciler.doTransitionStep(DbmsReconciler.java:335)
        at com.neo4j.dbms.DbmsReconciler.doTransitionStep(DbmsReconciler.java:336)
        at com.neo4j.dbms.DbmsReconciler.doTransitions(DbmsReconciler.java:318)
        at com.neo4j.dbms.DbmsReconciler.lambda$doTransitions$9(DbmsReconciler.java:308)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.ConcurrentModificationException
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
        at apoc.RegisterComponentFactory$RegisterComponentLifecycle.addResolver(RegisterComponentFactory.java:44)
        at apoc.ApocExtensionFactory$ApocLifecycle.lambda$init$0(ApocExtensionFactory.java:113)
        at java.base/java.util.HashMap$Values.forEach(HashMap.java:976)
        at apoc.ApocExtensionFactory$ApocLifecycle.lambda$init$1(ApocExtensionFactory.java:113)
        at apoc.ApocExtensionFactory$ApocLifecycle.withNonSystemDatabase(ApocExtensionFactory.java:89)
        at apoc.ApocExtensionFactory$ApocLifecycle.init(ApocExtensionFactory.java:95)
        at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.init(LifeSupport.java:405)
        ... 19 more
```

--- a/src/main/java/apoc/RegisterComponentFactory.java
+++ b/src/main/java/apoc/RegisterComponentFactory.java
@@ -11,7 +11,7 @@ import org.neo4j.logging.Log;
 import org.neo4j.logging.internal.LogService;
 import org.neo4j.procedure.impl.GlobalProceduresRegistry;
 
-import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.Map;
 
 public class RegisterComponentFactory extends ExtensionFactory<RegisterComponentFactory.Dependencies> {
@@ -38,10 +38,10 @@ public class RegisterComponentFactory extends ExtensionFactory<RegisterComponent
 
     public class RegisterComponentLifecycle implements Lifecycle {
 
-        private final Map<Class, Map<String, Object>> resolvers = new HashMap<>();
+        private final Map<Class, Map<String, Object>> resolvers = new ConcurrentHashMap<>();
 
         public void addResolver(String databaseNamme, Class clazz, Object instance) {
-            Map<String, Object> classInstanceMap = resolvers.computeIfAbsent(clazz, s -> new HashMap<>());
+            Map<String, Object> classInstanceMap = resolvers.computeIfAbsent(clazz, s -> new ConcurrentHashMap<>());
             classInstanceMap.put(databaseNamme, instance);
         }
 
@@ -52,9 +52,9 @@ public class RegisterComponentFactory extends ExtensionFactory<RegisterComponent
         @Override
         public void init() throws Exception {
             // FIXME: after lifecycle issue has been resolved upstream
-            resolvers.put(UuidHandler.class, new HashMap<>());
-            resolvers.put(TriggerHandler.class, new HashMap<>());
-            resolvers.put(CypherProceduresHandler.class, new HashMap<>());
+            resolvers.put(UuidHandler.class, new ConcurrentHashMap<>());
+            resolvers.put(TriggerHandler.class, new ConcurrentHashMap<>());
+            resolvers.put(CypherProceduresHandler.class, new ConcurrentHashMap<>());
             resolvers.forEach(
                     (clazz, dbFunctionMap) -> globalProceduresRegistry.registerComponent(clazz, context -> {
                         String databaseName = context.graphDatabaseAPI().databaseName();

